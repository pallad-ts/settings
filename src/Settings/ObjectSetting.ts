import {Maybe, Validation} from "monet";
import {Setting} from "../Setting";

export class ObjectSetting<TType extends Record<any, any>> extends Setting<TType> {
	async merge(values: Maybe<unknown[]>): Promise<Setting.Validator.Result<TType>> {
		if (values.isNone()) {
			return Validation.Success(this.defaultValue);
		}

		let currentValue = {
			...this.defaultValue
		};

		for (const override of values.some()) {
			const validationResult = await this.validator({
				...currentValue,
				...override as {}
			});
			if (validationResult.isFail()) {
				return validationResult;
			}
			currentValue = validationResult.success();
		}
		return Validation.Success(currentValue);
	}
}
