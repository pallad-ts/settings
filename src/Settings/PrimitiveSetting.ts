import {Maybe, Validation} from "monet";
import {Setting} from "../Setting";

export class PrimitiveSetting<TType> extends Setting<TType> {
	async merge(values: Maybe<unknown[]>): Promise<Setting.Validator.Result<TType>> {
		if (values.isNone()) {
			return Validation.Success(this.defaultValue);
		}

		const lastValue = values.some()[values.some().length - 1];
		return this.validator(lastValue);
	}
}
