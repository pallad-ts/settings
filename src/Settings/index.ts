export * from './ObjectSetting';
export * from './PrimitiveSetting';
