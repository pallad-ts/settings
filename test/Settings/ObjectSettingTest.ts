import {ObjectSetting} from "@src/Settings/ObjectSetting";
import {Maybe, Validation} from "monet";
import * as sinon from 'sinon';

describe('ObjectSetting', () => {
	const DEFAULT_VALUE = {
		foo: 'bar',
		zee: 'bee'
	};

	const DEFAULT_VALUE_COPY = JSON.parse(JSON.stringify(DEFAULT_VALUE));

	describe('merging', () => {
		const setting = new ObjectSetting(DEFAULT_VALUE);

		it('returns default values if no override provided', () => {
			return expect(setting.merge(Maybe.None()))
				.resolves
				.toEqual(Validation.Success(DEFAULT_VALUE));
		});

		it('fails if validation fails at any stage or merging', () => {
			const validator = sinon.stub().callsFake(value => {
				if (value.foo === 'test') {
					return Validation.Fail('Invalid value foo');
				}
				return Validation.Success(value);
			});

			const setting = new ObjectSetting(DEFAULT_VALUE, validator);

			return expect(setting.merge(Maybe.Some([{foo: 'test'}, {foo: 'bar'}])))
				.resolves
				.toMatchSnapshot();
		});

		it('returns value sanitized by validation', () => {
			const validator = sinon.stub().callsFake(value => {
				return Validation.Success(
					Object.fromEntries(
						Object.entries(value)
							.map(([key, value]) => [key, (value as string).toUpperCase()])
					)
				)
			});

			const setting = new ObjectSetting(DEFAULT_VALUE, validator);

			return expect(setting.merge(Maybe.Some([{foo: 'fee'}, {zee: 'boo'}])))
				.resolves
				.toEqual(Validation.Success({foo: 'FEE', zee: 'BOO'}));
		});

		afterEach(() => {
			expect(DEFAULT_VALUE).toEqual(DEFAULT_VALUE_COPY);
		});
	});
})
