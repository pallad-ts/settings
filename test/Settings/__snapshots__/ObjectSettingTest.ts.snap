// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectSetting merging fails if validation fails at any stage or merging 1`] = `
Object {
  "@@type": "monet.js/Validation",
  "isSuccessValue": false,
  "val": "Invalid value foo",
}
`;
